File: 0000-0999/0018.4Sum.go
func fourSum(nums []int, target int) [][]int {
    sort.Ints(nums)
    ans := [][]int{}
    kSum(nums, []int{}, target, 4, 0, &ans)
    return ans
}

func kSum(nums, path []int, target, k, lo int, ans *[][]int) {
    // O(n^(k-1))
    n := len(nums)
    if n-lo < k || k < 2 || target < k*nums[lo] || target > k*nums[n-1] {
        return
    }
    if k == 2 {
        hi := n - 1
        for lo < hi {
            if nums[lo]+nums[hi] == target {
                *ans = append(*ans, append(path, nums[lo], nums[hi]))
                for ; lo < hi && nums[lo] == nums[lo+1]; lo++ {}
                for ; lo < hi && nums[hi] == nums[hi-1]; hi-- {}
                lo, hi = lo+1, hi-1
            } else if nums[lo]+nums[hi] < target {
                lo++
            } else {
                hi--
            }
        }
    } else {
        for i := lo; i <= n-k; i++ {
            if i == lo || (i > lo && nums[i] != nums[i-1]) {
                kSum(nums, append(path, nums[i]), target-nums[i], k-1, i+1, ans)
            }
        }
    }
}

File: 0000-0999/0018.4Sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if len(nums) < 4:
            return []
        ans = []
        nums.sort()
        self.nSum(nums, 4, target, [], ans)
        return ans
        '''
        for i in xrange(len(nums)-3):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            for j in xrange(i+1, len(nums)-2):
                if j > i+1 and nums[j] == nums[j-1]:
                    continue
                l, r = j+1, len(nums)-1
                while l < r:
                    t = nums[i] + nums[j] + nums[l] + nums[r]
                    if t < target:
                        l += 1
                    elif t > target:
                        r -= 1
                    else:
                        ans.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l] == nums[l+1]:
                            l += 1
                        while l < r and nums[r] == nums[r-1]:
                            r -= 1
                        l += 1
                        r -= 1
        return ans
        '''
    def twoSum(self, nums, target, temp, ans):
        l, r = 0, len(nums)-1
        while l < r:
            t = nums[l] + nums[r]
            if t < target:
                l += 1
            elif t > target:
                r -= 1
            else:
                ans.append(temp + [nums[l], nums[r]])
                while l < r and nums[l] == nums[l+1]:
                    l += 1
                while l < r and nums[r] == nums[r-1]:
                    r -= 1
                l += 1
                r -= 1
        return
        
    def nSum(self, nums, N, target, temp, ans):
        if N < 2 or len(nums) < N:
            return
        if N == 2:
            self.twoSum(nums, target, temp, ans)
        else:
            if target > N*nums[-1]:
                return
            for i in xrange(len(nums)-N+1):
                if target < N*nums[i]:
                    break
                if i > 0 and nums[i] == nums[i-1]:
                    continue
                self.nSum(nums[i+1:], N-1, target-nums[i], temp+[nums[i]], ans)
        return
    